execute/group_lookup.go

```diff
@@ -227,8 +227,21 @@ func (l *GroupLookup) createOrSetInGroup(index int, key flux.GroupKey, value int
 	// Construct the new group entry and copy the end of the slice
 	// into the new key group.
 	l.groups[index+1] = func() *groupKeyList {
+		// TODO(rockstar): A nice optimization here would be to prevent
+		// the deleted items from being copied. However, this entire function
+		// needs to be refactored to support that, as it's possible that *all*
+		// the elements have been deleted, so no split is needed.
+		// Moving currently deleted elements out of this key group, the deleted
+		// count must be decremented.
+		for _, item := range kg.elements[i:] {
+			if item.deleted {
+				kg.deleted--
+			}
+		}
```

Comment:
If I understand correctly a key group is deleted once its deleted count is equal to the number of elements in its list. See https://github.com/influxdata/flux/pull/2726/files#diff-58e8c89e9ef3bf1dc7572bfd10e03d34R286 

So during an unrelated Delete operation if a key group contains only deleted items it will be removed entirely.

Because we were copying deleted items out of a key group and not decrementing the deleted count, a bug could arise where items that should not be deleted were because the deleted count of the key group was equal to the number of elements in the group. 

That would explain the behavior we were seeing in testing.diff where groups seems to disappear for no reason. 

Good find!

Have you tested `testing.diff` with this fix to validate it fixes it? I know the RandomGroupLookup fixed the issue already, I am curious is this fix would have also fixed it as well.

Reply:
I did check it against `testing.Diff`, but `testing.Diff` is probably better served with the changes in #2721

Reply:
Awesome, I agree, the RandomAccess implementation is the appropriate one for testing.diff. I am also glad to hear that this fix closes the loop on all the testing.diff weirdness.